src/modm/driver/encoder/encoder_input.hpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_ENCODER_HPP
+#define MODM_ENCODER_HPP
+
+#include <modm/platform.hpp>
+#include <bit>
+#include <type_traits>
+
+namespace modm
+{
+/**
+ * \brief	Encoder
+ *
+ * \ingroup	modm_ui_encoder
+ * \author	2021 Thomas Sommer
+ */
+
+/**
+ * @ingroup modm_driver_encoder_input
+ * @author  Thomas Sommer
+ *
+ * @brief					This driver decodes a AB (incremental) encoder signal
+ *
+ * @tparam SignalA          	First modm::platform::Gpio pin to input the encoder signal.
+ * @tparam SignalB          	Second modm::platform::Gpio pin to input the encoder signal.
+ * @tparam DIVISION		 	Resolution: count of phasechange for one increment.
+ * @tparam UnderlyingType	Must be signed integer
+ */
+template<
+	typename SignalA,
+	typename SignalB,
+	uint8_t DIVISION = 4,
+	std::signed_integral UnderlyingType = int8_t
+	>
```

Comment:
Im busy with another thing right now.
Later, i definitely add a static_cast and check, if the size of UnderlyingType can at least handle DIVISION